### ISSUE TITLE
docker sees all GPUs despite access configured via Slurm (cgroups)

### ISSUE THREAD

**Issue by mi-lad** [NVIDIA/nvidia-docker#1227]
We want to use nvidia-docker inside our slurm jobs. Inside the slurm job only the GPUs we have requested are visible when we look at `nvidia-smi` outputs. I was expecting if I do `docker run --gpus all nvidia-smi` I would also see the same number GPUs. However within the container nvidia-smi shows _all_ gpus available on that machine. The container seems to have the correct `CUDA_VISIBLE_DEVICES` environment variable but not the correct permissions.

Could it be related to the fact that by default docker needs sudo and we've created a user group so that we don't need sudo?

(BTW I noticed some things are different in Fedora but we're using Ubuntu.)

**Comment by flx42** [comment]
By default, docker containers are part of the Docker daemon's cgroups (since they are spawned by the daemon that is not a child of your job allocation). So the device cgroups of your job is not being applied.

**Comment by trn84** [comment]
Is there anything planned on this?

**Comment by flx42** [comment]
@trn84 not sure what *could* be done, that's the design of Docker. You can try to play around with `--cgroup-parent`, or spawn an unprivileged docker daemon in the job context, but that's about it.

You can look at this page for an overview of options:
https://slurm.schedmd.com/containers.html

**Comment by trn84** [comment]
yes, I was thinking about the cgroup-parent coming from the slurm context. There is a tool called socker that was going in that direction but did not work for me.

I am using slurm+enroot+pyxis which seems to work but is kinda a hassle with the additional container infrastructure...

**Comment by elgalu** [comment]
@trn84 what was your experience these months with slurm+enroot+pyxis so far?

**Comment by rkindi** [comment]
Not sure if it's relevant since I am a slurm and docker noob, but just wanted to share what worked for my use case. I'm not using nvidia-docker, but am using docker for some gpu work and was able to use the following:

`# docker run -it --gpus $(scontrol show job -d $SLURM_JOB_ID | grep -oP '(?<=GRES=gpu:1\(IDX:)[0-9]+(?=\))') my_image:latest`

The command gets the physical id of the gpu (I only request 1 gpu for my use case) using `scontrol show job -d` and passes it into the `docker run` command. Of course, this command can be made more generic if you are using multiple gpus.

**Comment by adamnovak** [comment]
I tried `--gpus=\"device=$SLURM_STEP_GPUS\"` as suggested at https://docs.docker.com/config/containers/resource_constraints/#expose-gpus-for-use and that also seemed to work.

It would be great if Slurm and Docker and the nvidia runtime developed some way to pass this information around behind the scenes, or at least made it easy and well-documented for users to make the necessary connection. Maybe a "${SLURM_DOCKER_FLAGS[@]}" could be provided by Slurm, ready to be added to a `docker run` command?
